backend/internal/consumer/storage_updater.go

```diff
@@ -0,0 +1,330 @@
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/ducanhpham0312/zeevision/backend/internal/storage"
+)
+
+const (
+	// Maximum attempts to try creating eg. a variable when it might be
+	// failing due to the instance not existing yet (race handling
+	// nigh-simultaneous incoming records)
+	maxCreateAttempts  = 5
+	createAttemptDelay = 100 * time.Millisecond // nolint:gomnd
+)
+
+type storageUpdater struct {
+	storer storage.Storer
+
+	msgChannel   msgChannelType
+	closeChannel signalChannelType
+
+	wg *sync.WaitGroup
+}
+
+func newDatabaseUpdater(storer storage.Storer, msgChannel msgChannelType, closeChannel signalChannelType, wg *sync.WaitGroup) *storageUpdater {
+	result := &storageUpdater{
+		storer: storer,
+
+		msgChannel:   msgChannel,
+		closeChannel: closeChannel,
+		wg:           wg,
+	}
+
+	// launch a database handler goroutine
+	// Have a pool of five so we can (hopefully) avoid data races by retrying
+	poolSize := 5
+	for i := 0; i < poolSize; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			result.storageUpdaterLoop()
+		}()
+	}
+
+	return result
+}
+
+// Handle actual database updates from consumers
+func (u *storageUpdater) storageUpdaterLoop() {
+	closeChannel := u.closeChannel
+	msgChannel := u.msgChannel
+readLoop:
+	for {
+		select {
+		case <-closeChannel:
+			// Close this one too when we get a closeChannel message
+			break readLoop
+		case msg := <-msgChannel:
+			var untypedRecord UntypedRecord
+			err := json.Unmarshal(msg, &untypedRecord)
+			if err != nil {
+				log.Printf("Failed to unmarshal: %v", err)
+				continue readLoop
+			}
+
+			err = u.handlingDispatch(&untypedRecord)
+			if err != nil {
+				log.Printf("Handling failed: %v", err)
+				continue readLoop
+			}
+		}
+	}
+}
+
+func (u *storageUpdater) handlingDispatch(untypedRecord *UntypedRecord) error {
+	var err error
+
+	switch untypedRecord.ValueType { // nolint:exhaustive
+	case "":
+		return fmt.Errorf("zero-value value type in record")
+	case ValueTypeDeployment:
+		err = u.handleDeployment(untypedRecord)
+		if err != nil {
+			return fmt.Errorf("failed to handle deployment: %w", err)
+		}
+	case ValueTypeProcess:
+		err = u.handleProcess(untypedRecord)
+		if err != nil {
+			return fmt.Errorf("failed to handle process: %w", err)
+		}
+	case ValueTypeProcessInstance:
+		err = u.handleProcessInstance(untypedRecord)
+		if err != nil {
+			return fmt.Errorf("failed to handle process instance: %w", err)
+		}
+	case ValueTypeVariable:
+		err = u.handleVariable(untypedRecord)
+		if err != nil {
+			return fmt.Errorf("failed to handle variable: %w", err)
+		}
+	default:
+		log.Printf("Unhandled record: %v (intent: %v)",
+			untypedRecord.ValueType, untypedRecord.Intent)
+	}
+	return nil
+}
+
+func (u *storageUpdater) handleDeployment(untypedRecord *UntypedRecord) error {
+	storer := u.storer
+
+	record, err := WithTypedValue[DeploymentValue](*untypedRecord)
+	if err != nil {
+		return fmt.Errorf("failed to cast: %w", err)
+	}
```

Comment:
I see your point, but I just don't like the semantics of "here is a value you may not be able to use". My argument was more about the correctness and reducing possibility for invalid usage, rather than short and easy to read code. If Go had some good abstractions over error handling, maybe we wouldn't even have this discussion.

But this is fine to leave as is, because it works and will possibly be overridden by topic-based handling.